pkg/proxy/healthcheck/service_health.go

```diff
@@ -75,14 +85,20 @@ func newServiceHealthServer(hostname string, recorder events.EventRecorder, list
 		recorder:      recorder,
 		listener:      listener,
 		httpFactory:   factory,
+		nodeState:     nodeState{},
+		healthzServer: healthzServer,
 		services:      map[types.NamespacedName]*hcInstance{},
 		nodeAddresses: nodeAddresses,
 	}
 }
 
 // NewServiceHealthServer allocates a new service healthcheck server manager
-func NewServiceHealthServer(hostname string, recorder events.EventRecorder, nodePortAddresses []string) ServiceHealthServer {
-	return newServiceHealthServer(hostname, recorder, stdNetListener{}, stdHTTPServerFactory{}, nodePortAddresses)
+func NewServiceHealthServer(hostname string, recorder events.EventRecorder, nodePortAddresses []string, healthzServer ProxierHealthUpdater) ServiceHealthServer {
+	return newServiceHealthServer(hostname, recorder, stdNetListener{}, stdHTTPServerFactory{}, nodePortAddresses, healthzServer)
+}
+
+type nodeState struct {
+	isHealthy bool
```

Comment:
don't we have to do this variable thread safe, it is accessed from multiple places
use atomic perhaps?

Reply:
Good point, we probably need to "wrap" `SyncNode` in 

```
hcs.lock.Lock()
defer hcs.lock.Unlock()
```

As has been done for `SyncService`. The read should already be protected though in `ServeHTTP` by the read lock.

Reply:
I like this atomic bool vars for these cases, your choice though

https://github.com/kubernetes/kubernetes/blob/ee505677f70439ff8f5515cc899eab7c18cdb1dd/pkg/proxy/iptables/proxier.go#L588-L598

https://gist.github.com/mahan/6256149

Reply:
But in this case there really is only one writer, there's no concurrency. `SyncNode` is only invoked by the node informer, and its event handlers (add/update/delete) cannot be called concurrently. So we hence only ever have one writer at any given moment in time. 

Making it an atomic bool would give the impression to people reading the code that we are dealing with concurrency, where in fact that is not the case.

Reply:
Nevermind what I said above, I was confused. There is concurrency, but I find the code reads better by using `hcs.lock`